lighthouse-core/gather/computed/request-chains.js

```diff
@@ -0,0 +1,124 @@
+   * @return {{request: !Object, children: !Object}}
+   */
+  static extractChain(networkRecords, predicate) {
+    predicate = predicate || (() => true);
+    networkRecords = networkRecords.filter(req => req.finished);
+
+    // Build a map of requestID -> Node.
+    const requestIdToRequests = new Map();
+    for (const request of networkRecords) {
+      requestIdToRequests.set(request.requestId, request);
+    }
+
+    const requestsToExamine = networkRecords.filter(predicate);
+
+    // Create a tree of critical requests.
+    const requestChains = {};
+    for (const request of requestsToExamine) {
+      // Work back from this request up to the root. If by some weird quirk we are giving request D
+      // here, which has ancestors C, B and A (where A is the root), we will build array [C, B, A]
+      // during this phase.
+      const ancestors = [];
+      let ancestorRequest = request.initiatorRequest();
+      let node = requestChains;
+      while (ancestorRequest) {
+        // If the parent request isn't a high priority request it won't be in the
+        // requestIdToRequests map, and so we can break the chain here. We should also
+        // break it if we've seen this request before because this is some kind of circular
+        // reference, and that's bad.
+        if (!predicate(ancestorRequest) || ancestors.includes(ancestorRequest.requestId)) {
+          // Set the ancestors to an empty array and unset node so that we don't add
+          // the request in to the tree.
+          ancestors.length = 0;
+          node = undefined;
+          break;
+        }
+        ancestors.push(ancestorRequest.requestId);
+        ancestorRequest = ancestorRequest.initiatorRequest();
+      }
+
+      // With the above array we can work from back to front, i.e. A, B, C, and during this process
+      // we can build out the tree for any nodes that have yet to be created.
+      let ancestor = ancestors.pop();
+      while (ancestor) {
+        const parentRequest = requestIdToRequests.get(ancestor);
+        const parentRequestId = parentRequest.requestId;
+        if (!node[parentRequestId]) {
+          node[parentRequestId] = {
+            request: RequestChains.flattenRecord(parentRequest),
+            children: {}
+          };
+        }
+
+        // Step to the next iteration.
+        ancestor = ancestors.pop();
+        node = node[parentRequestId].children;
+      }
+
+      if (!node) {
+        continue;
+      }
+
+      // If the node already exists, bail.
+      if (node[request.requestId]) {
+        continue;
+      }
+
+      // node should now point to the immediate parent for this request.
+      node[request.requestId] = {
+        request: RequestChains.flattenRecord(request),
+        children: {}
+      };
+    }
+
+    return requestChains;
+  }
+
+  /**
+   * @param {!DevtoolsLog} devtoolsLog
+   * @param {!ComputedArtifacts} artifacts
+   * @return {!Promise<!Object>}
```

Comment:
oh, just kidding, we did it in [`crc-details-renderer`](https://github.com/GoogleChrome/lighthouse/blob/57bcf43a661199d55a7f39ecc008815a2da1fcc6/lighthouse-core/report/v2/renderer/crc-details-renderer.js#L203). Can use `CriticalRequestChainRenderer.CRCNode` but we can make it a generic (not just critical) RequestNode type?